### PR TITLE
First CashTokens support in CashScript SDK

### PR DESCRIPTION
**Pull request by mr-zwets**

### CODE REVIEW

**Review by rkalis — CHANGES_REQUESTED** [review]
Now that Libauth v2 has been integrated, I'm guessing that we'll also need to add a few things about P2SH32 and related updates.

File: packages/cashscript/src/Errors.ts [hunk 1]
```diff
@@ -8,8 +7,14 @@ export class TypeError extends Error {
...
 
 export class OutputSatoshisTooSmallError extends Error {
-  constructor(satoshis: bigint) {
-    super(`Tried to add an output with ${satoshis} satoshis, which is less than the DUST limit (${DUST_LIMIT})`);
+  constructor(satoshis: bigint, minimumAmount:bigint) {
+    super(`Tried to add an output with ${satoshis} satoshis, which is less than the required minimum for this output-type (${minimumAmount})`);
+  }
+}
+
+export class TokensToNonTokenAddress extends Error {
```
[the remark this answers is not in the document]
Comment (mr-zwets) [reply packages/cashscript/src/Errors.ts:15]: sure

File: packages/cashscript/src/Transaction.ts [hunk 2]
```diff
@@ -91,12 +93,15 @@ export class Transaction {
...
-  to(to: string, amount: bigint): this;
+  to(to: string, amount: bigint, tokenDetails: TokenDetails): this;
   to(outputs: Recipient[]): this;
 
-  to(toOrOutputs: string | Recipient[], amount?: bigint): this {
+  to(toOrOutputs: string | Recipient[], amount?: bigint, tokenDetails?: TokenDetails): this {
     if (typeof toOrOutputs === 'string' && typeof amount === 'bigint') {
-      return this.to([{ to: toOrOutputs, amount }]);
+      const recipient = { to: toOrOutputs, amount, token: tokenDetails };
+      if(typeof tokenDetails != "undefined") recipient.token = tokenDetails;
```
[the remark this answers is not in the document]
Comment (mr-zwets) [reply packages/cashscript/src/Transaction.ts:100]: ah thanks for catching this!

Not sure which is better to have the key regardless or only when it actually contains a token object.
I made the token key optional in the types for `Recipient` & `Output`
Relevant for this code for example
```
if ('token' in recipient) {
    if(!isTokenAddress(recipient.to)){
      throw new TokensToNonTokenAddress(recipient.to);
    }
  }
```
can change it to `if(recipient.token){` if the value is undefined for non token holding recipients...

File: packages/cashscript/src/Transaction.ts [hunk 3]
```diff
@@ -91,12 +93,15 @@ export class Transaction {
...
+      validateRecipient(recipient);
```
[the remark this answers is not in the document]
Comment (mr-zwets) [reply packages/cashscript/src/Transaction.ts:101]: ah makes sense

File: packages/cashscript/src/Transaction.ts [hunk 4]
```diff
@@ -273,6 +283,112 @@ export class Transaction {
       throw Error('Attempted to build a transaction without outputs');
     }
 
+    // Construct object with total output of fungible tokens by tokenId
+    const netBalanceTokens: any = {};
```
[the remark this answers is not in the document]
Comment (mr-zwets) [reply packages/cashscript/src/Transaction.ts:284]: awesome thanks! I knew type any wasn't good practice but didn't know this is the correct approach!

File: packages/cashscript/src/Transaction.ts [hunk 5]
```diff
@@ -273,6 +283,112 @@ export class Transaction {
...
+      } else {
+        netBalanceTokens[tokenCategory] -= output.token.amount;
+      }
+      if(typeof output.token.nft !== "undefined"){
+        listNftsOutputs.push({...output.token.nft, category: output.token.category});
+      }
+    }
+    // If inputs are manually provided, check token balances
+    if(this.inputs.length > 0){
+      for(const tokenBalance of netBalanceTokens){
```
[the remark this answers is not in the document]
Comment (mr-zwets) [reply packages/cashscript/src/Transaction.ts:317]: right 👍 could also keep the for(... of ...) with
`for (const [category, balance] of Object.entries(netBalanceTokens)) {`

File: packages/cashscript/src/Transaction.ts [hunk 6]
```diff
@@ -273,6 +283,112 @@ export class Transaction {
...
+        // Add token change outputs if applicable
+        if(this.tokenChange && tokenBalance.amount > 0){
+          const tokenDetails: TokenDetails = {
+            category : tokenBalance.category,
+            amount : BigInt(-tokenBalance.amount)
```
[the remark this answers is not in the document]
Comment (mr-zwets) [reply packages/cashscript/src/Transaction.ts:322]: I think I added the minus sign when the object used to hold output balances and subtracted inputs so change would be a negative number. So it's there still by mistake 😅

File: packages/cashscript/src/Transaction.ts [hunk 7]
```diff
@@ -273,6 +283,112 @@ export class Transaction {
...
+          };
+          const tokenChangeOutput = { to: this.address, amount: BigInt(1000), token: tokenDetails };
+          this.outputs.push(tokenChangeOutput);
+        }
+        // Throw error when token balance is insufficient
+        if(tokenBalance.amount < 0){
+          throw new Error(`Insufficient token balance.`);
+        }
+      }
+      // Compare nfts in- and outputs, check if inputs have nfts corresponsing to outputs
```
[the remark this answers is not in the document]
Comment (mr-zwets) [reply packages/cashscript/src/Transaction.ts:332]: > If you could give this point some more thought that would be much appreciated. In any case I think we'll need to add some really good edge case tests for this to make sure things are working as expected.

Sure will give it some more thought. Do you agree on the high level that it's correct to keep a  `netBalanceTokens` object for fungible tokens with tokenId as key and amount as value (where inputs are added to the amount and outputs subtracted). And that for NFTs a list of outputs is kept and checked against a list of NFT inputs?

I agree edge case testing is really important for this part of the code

File: packages/cashscript/src/Transaction.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkalis) [reply packages/cashscript/src/Transaction.ts:100]: I think it's probably fine to have the key regardless (and just have the value be undefined). And then like you said, we should use `if (recipient.token) { ... }`.

File: packages/cashscript/src/Transaction.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rkalis) [reply packages/cashscript/src/Transaction.ts:317]: Yeah you can also keep the `for ... of` in that case 👌

File: packages/cashscript/src/Transaction.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rkalis) [reply packages/cashscript/src/Transaction.ts:332]: Yeah I think overall that kind of approach seems like the way to go 👌

File: packages/cashscript/src/Transaction.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rkalis) [reply packages/cashscript/src/Transaction.ts]: I just checked out the code locally, and linting rules state that using `continue` is forbidden. I generally try to keep the eslint config pretty close to the airbnb styleguide config (can be checked in `.eslintrc.cjs`). 

In cases like this I don't see a great reason for why we'd need to make an exception here, even though I'm not sure about the rationale behind airbnb's decision to block `continue` statements 🤷‍♂️

So in that case we need to replace the `for ... of` loops with `array.forEach(() => {})` and use `return` statements where you use `continue` statements now. Most other linting issues can just get fixed by running `yarn lint --fix` in the separate packages.

**Comment by rkalis** [comment]
There are still some open linting issues and we still have to add some more tests. But we'll do that before merging `next` into `master`.
